numpy/_core/src/multiarray/ctors.c

```diff
@@ -720,7 +720,46 @@ PyArray_NewFromDescr_int(
         }
     }
 
-    fa = (PyArrayObject_fields *) subtype->tp_alloc(subtype, 0);
+    npy_bool data_on_instance = NPY_FALSE;
+    if (subtype == &PyArray_Type) {
+        /*
+         * For an array scalar that is not too big, we allocate the required
+         * memory as part of the object, to avoid the overhead of allocating
+         * tracked memory with PyDataMem_UserNew.
+         *
+         * We do this bypassing the standard tp_alloc. In principle, on python
+         * 3.13 it seemed possible not to, as long as one uses a PyVarObject
+         * and tp_itemsize=1, but on python 3.11 this gave problems with
+         * subclassing.  An advantage of the present approach is that it does
+         * not change the size of a regular instance.  Note that we mimic
+         * PyType_GenericAlloc here in zeroing the bytes (python 3.13 zeros
+         * only the extra ones, 3.11 all of them), so that does not have to
+         * be done later.
+         *
+         * TODO: maybe do this generally for any small allocation, i.e.,
+         * calculate full nbytes up front and just use that here?
+         *
+         * Note: alignment constraint written such that compiler can just
+         * show it is true.
```

Comment:
Makes sense, but since this is a static  check, I think it should be a `static_assert` (otherwise you won't notice if it starts failing).  Might also use `NPY_ALIGNOF(npy_clongdouble)`, not sure that we can actually trust `sizeof(npy_intp)` to be a maximally aligned type.

Reply:
Actually, not sure about `NPY_ALIGNOF(npy_clongdouble)` - if that can be more than `sizeof(npy_intp)`, then I'm not sure we can count on `PyObject_Malloc` to allocate to that level (i.e., our basic size may be 96 and thus divisible by 16, but malloc may only care about alignment of 8?). If that's a problem, it might have to be a check on `data` itself... 

But probably it is not a problem. How about,
```
            static_assert((sizeof(PyArrayObject_fields)
                           & (sizeof(npy_intp) - 1)) == 0);
            static_assert(NPY_ALIGNOF(npy_clongdouble) <= sizeof(npy_intp));
```

Reply:
Any malloc implementation (including Python's) guarantees at least the maximum alignment of any type on the system (Not necessarily the maximum of any instruction, but that is a different matter).
If that assertion fails, that is OK. Should have just written the previous one with `NPY_ALIGNOF(npy_clongdouble)` instead (might as well use `% == 0` since speed is irrelevant).

If that _also_ fails, then that is also OK! Just introduce a helper struct to clarify this.  Maybe you prefer that either way even...
```
struct {
    PyArrayObject_fields arr;
    npy_clongdouble data[];
} _sizeof_arr_plus_data_struct;
```
then you can skip on the asssertions and use the sizeof that.

Reply:
Yes, we could.  But no need to "remove" anything, the only thing for the removal is that it uses `data[1]` rather than `data[]`.  And the only reason for that is that the `data[]` notation is C99 (and I am not sure which C++ versios support it).
Hopefully C++17 does, because otherwise we may have to work around things.  But we do use `data[]` a lot in NumPy, I am just not sure that it is in public headers yet (and this has to be public).

Reply:
OK, I went ahead and added `_storage`, with an explanation. Also added a `_PyArray_DATA_ON_INSTANCE` function for checking whether that is the case. It makes the other code changes more obvious. Not sure it should be in `ndarraytypes.h`, but didn't know where else to put it (and there are other private ones in there).